### ISSUE TITLE
`nonminimal_bool` reorders operands, changing behavior

### ISSUE THREAD

**Issue by khyperia** [rust-lang/rust-clippy#8050]
### Summary

In some cases (not sure if this is an actual minimal repro or if it's more complicated than needed, I don't really understand what's going on here), `nonminimal_bool` will reorder the operands to logical && and || operators, changing behavior if the operands have side effects. I hit this in real-world code, and got lucky that I noticed it after quickly applying clippy's fix - my code was of the form `ok_to_insert() && hash_set.insert(...)`, which would be really bad if flipped.

### Reproducer

I tried this code:

```rust
fn f() -> bool {
    println!("f() called");
    false
}

fn main() {
    let a = 5;
    let _ = !(a == 2 && !f() || a == 3 && !f());
}
```

I expected to see a suggestion to rewrite to the following:

```rust
let _ = (a != 2 || f()) && (a != 3 || f());
```

Instead, this happened:

```rust
warning: this boolean expression can be simplified
 --> src/main.rs:8:13
  |
8 |     let _ = !(a == 2 && !f() || a == 3 && !f());
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `(f() || a != 3) && (f() || a != 2)`
  |
  = note: `#[warn(clippy::nonminimal_bool)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#nonminimal_bool
```

This changes the behavior of the program, due to changing the order of short-circuiting operators: the program now prints `f() called` twice, whereas originally it didn't print anything at all.

### Version

```text
rustc 1.56.1 (59eed8a2a 2021-11-01)
binary: rustc
commit-hash: 59eed8a2aac0230a8b53e89d4e99d55912ba6b35
commit-date: 2021-11-01
host: x86_64-unknown-linux-gnu
release: 1.56.1
LLVM version: 13.0.0
```

**Comment by llogiq** [comment]
So the solution would be to avoid linting on detecting a combination of short-circuiting ops and (potential) side effects?

**Comment by khyperia** [comment]
> So the solution would be to avoid linting

Hm? No, as described in the "reproducer"/"I expected to see" section, the solution would be to rewrite to an expression that does not change the behavior of the program, rather than one that does.

**Comment by surechen** [comment]
Hi, I’m  interested in this issue and I’m trying to find out why this happened.

**Comment by surechen** [comment]
I think because the `simplify` function in crate `quine-mc_cluskey` flipped the values on the left and right sides of `||` when processing !((a&&b)||(c&&d)). For some cases SuggestContext::recurse can flip it back , But for some mixed situation of multiple || and &&, it doesn't seem to work as expected. I don't konw how to fix at present.
